src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -206,32 +252,58 @@ static SyntaxNode TryGetAcceptedNodeOrExtracted(SyntaxNode node, Predicate<Synta
                     return node;
                 }
 
-                var extractedNode = extractNode(node, syntaxFacts);
+                var extractedNode = extractNode(node, syntaxFacts, extractParentsOfHeader);
                 return (extractedNode != null && predicate(extractedNode))
                     ? extractedNode
                     : null;
             }
         }
 
+        private static TextSpan GetSpanWithoutAttributes(SyntaxNode node, SyntaxNode root, ISyntaxFactsService syntaxFacts)
+        {
+            // Span without AttributeLists
+            // - No AttributeLists -> original .Span
+            // - Some AttributeLists -> (first non-trivia/comment Token.Span.Begin, original.Span.End)
+            //   - We need to be mindful about comments due to:
+            //      // [Test1]
+            //      //Comment1
+            //      [||]object Property1 { get; set; }
+            //     the comment node being part of the next token's (`object`) leading trivia and not the AttributeList's node.
+            var attributeList = syntaxFacts.GetAttributeLists(node);
+            if (attributeList.Any())
+            {
+                var endOfAttributeLists = attributeList.Last().Span.End;
+                var afterAttributesToken = root.FindTokenOnRightOfPosition(endOfAttributeLists);
+
+                return TextSpan.FromBounds(afterAttributesToken.Span.Start, node.Span.End);
+            }
+
+            return node.Span;
+        }
```

Comment:
fwiw all this code is looking great.  It's *getting* to the point though that i think it would be sensible to have a *dedicated* set of tests around this service (similar to how we test ITypeInferenceService directly).  Testing through individual features is also great.  But it might be really nice to have just tests of this service so we can target things directly.  This will be esp. true if you start adding the concepts of 'headers' and whatnot.  Instead of needing to update the tests in 50 features, you can add the tests (in nicely separated regions/files) for all the different types of logic here.

Reply:
Yep, that's actually what I was thinking as well. The issue is that I'm not entirely sure how to go around testing it (without creating all infrastructure from the beginning). Is there any similar service (location within text file -> SyntaxNode) I could look at for inspiration?

Reply:
It would be really simple.  Look at how the TypeInferrerTests work.  All you need is a single helper function that basically is given the markup text, and likely the type of node you're asking is selected.  You'd want an overload saying "this should succeed" and one saying "this should fail".  If you want to ensure the precise node is selected, your markup can specify both the selection-span and a named span for the resultant node.  so it would look like this:

```c#
TestAsync<PropertyDeclarationSyntax>(
@"
class C
{
    public int [|Foo|]() { }
}");
```

or


```c#
TestAsync(
@"
class C
{
    {|result:public int [|Foo|]() { }|}
}");
```

And so on and so forth.  All it will do internally is make a workspace, add a document, get your service, call into it with the right values, and then check the results against what the test says it wants.

Reply:
Where should the tests go to? The same as TypeInferr tests? I.e. base to `EditorFeatures/Test.Uitilities/RefactoringHelpers/`, … and tests themselves `EditorFeatures/CSharp.EditorFeatures.UnitTests/RefactoringHelpers`

Reply:
yes (and the vb equivalent), that seems reasonable :)